python/pyspark/sql/dataframe.py

```diff
@@ -79,11 +81,18 @@ def __init__(self, jdf, sql_ctx):
     @property
     @since(1.3)
     def rdd(self):
-        """Returns the content as an :class:`pyspark.RDD` of :class:`Row`.
+        """Returns the content as an :class:`pyspark.RDD` of :class:`Row` or custom object.
         """
         if self._lazy_rdd is None:
             jrdd = self._jdf.javaToPython()
-            self._lazy_rdd = RDD(jrdd, self.sql_ctx._sc, BatchedSerializer(PickleSerializer()))
+            if self._jdf.isOutputPickled():
+                # If the underlying java DataFrame's output is pickled, which means the query
+                # engine don't know the real schema of the data and just keep the pickled binary
+                # for each custom object(no batch).  So we need to use non-batched serializer here.
+                deserializer = PickleSerializer()
```

Comment:
The overhead of PickleSerializer is pretty high, it will serialize the class for each row, could you do some benchmark to see how is the difference between non-batched vs batched (both size and CPU time)?